airbyte-ci/connectors/pipelines/pipelines/gradle.py

```diff
@@ -121,29 +122,69 @@ async def _run(self) -> StepResult:
+            # Update the persistent gradle cache by resolving all dependencies.
+            # The idea here is to have this persistent cache contain little more than jars and poms.
+            # Also, build the java CDK and publish it to the local maven repository.
+            .with_exec(
+                sh_dash_c(
+                    [
+                        # Ensure that the local maven repository root directory exists.
+                        "mkdir -p /root/.m2",
+                        # Load from the persistent cache.
+                        "(rsync -a --stats /root/gradle-persistent-cache/ /root/.gradle || true)",
+                        # Resolve all dependencies and write their checksums to './gradle/verification-metadata.dryrun.xml'.
+                        f"./gradlew {' '.join(self.DEFAULT_GRADLE_TASK_OPTIONS)} --write-verification-metadata sha256 help --dry-run",
+                        # Store to the persistent cache.
+                        "(rsync -a --stats /root/.gradle/ /root/gradle-persistent-cache || true)",
+                        # Build the CDK and publish it to the local maven repository.
+                        # Do this last to not pollute the persistent cache.
+                        f"./gradlew {' '.join(self.DEFAULT_GRADLE_TASK_OPTIONS)} :airbyte-cdk:java:airbyte-cdk:publishSnapshotIfNeeded",
+                    ]
+                )
+            )
+        )
+
+        # Mount only the code needed to build the connector.
+        # This reduces the scope of the inputs to help dagger reuse container layers.
+        # The contents of '/root/.gradle' and '/root/.m2' are by design not overly sensitive to changes in the rest of the git repo.
+        gradle_container = (
+            gradle_container_base
             # TODO: remove this once we finish the project to boost source-postgres CI performance.
             .with_env_variable("CACHEBUSTER", hacks.get_cachebuster(self.context, self.logger))
-            # Mount the gradle cache volume.
-            # We deliberately don't mount it at $GRADLE_HOME, instead we load it there and store it from there using rsync.
-            # This is because the volume is accessed concurrently by all GradleTask instances.
-            # Hence, why we synchronize the writes by setting the `sharing` parameter to LOCKED.
-            .with_mounted_cache("/root/gradle-cache", self.connector_java_build_cache, sharing=CacheSharingMode.LOCKED)
-            # Mount the parts of the repo which interest us in /airbyte.
-            .with_workdir("/airbyte")
+            # Mount the connector-agnostic whitelisted files in the git repo.
             .with_mounted_directory("/airbyte", self.context.get_repo_dir(".", include=include))
+            # Mount the sources for the connector and its dependencies.
             .with_mounted_directory(str(self.context.connector.code_directory), await self.context.get_connector_dir())
```

Comment:
Don't we already have the connector code at this stage as we already mounted the full repo in the last `with_mounted_directory` operation?

Reply:
No, we deliberately don't base this container off the container with the whole repo mounted. 

The reason for this is that mounting the whole repo is only necessary to, more or less, download all the jars and compile the CDK. These outputs change far less often than the git repo itself.

Consider the following scenario: you create a PR up which makes changes to two unrelated connectors. We'll have airbyte-ci run and the following will happen:
- All the layers in `with_whole_git_repo` need to be rebuilt; that's normal, the git repo is different than on master.
- All the layers in `gradle_container` also need to be rebuilt, as expected.
 
Now, let's say you push a commit which only changes one of the two connectors and `airbyte-ci` runs on the same runner:
- All the layers in `with_whole_git_repo` need to be rebuilt; that's normal, the git repo is different again.
- The resulting `/root/.gradle` and `/root/.m2` directors are rebuilt but they end up being the same as in the previous airbyte-ci run, because there were no changes to the dependencies and to the CDK sources.
- For the connector which was not changed, the layers in `gradle_container` can be re-used.

What this means is that the layers for the connector container (with the partial git repo mount) are more likely to be reused.

Reply:
To be clear: this is mostly speculation on my part and this deserves a test. I'll try this out on my machine to confirm!

Reply:
As it turns out, this speculation was wrong. I rewrote this a half-dozen times and still couldn't get `/root/.gradle` to be reused, even when paring it down to `/root/.gradle/caches/modules-2` as explained in https://docs.gradle.org/current/userguide/dependency_resolution.html#sub:cache_copy and below.

Reply:
I was able to get this to work in the end, but it required mounting connector sources from the host instead of from the other container!